ios/Classes/ComWilliamrijksenOnesignalModule.m

```diff
@@ -76,35 +87,141 @@ - (void) actionHandler:(OSNotificationOpenedResult *)result {
+                                    NSLog(@"[DEBUG] com.williamrijksen.onesignal User opened notification");
+                                    [self actionHandler:result];
+                                }
+                            settings:@{kOSSettingsKeyAutoPrompt: @false}
+    ];
+}
+
++ (void)load
+{
+    NSLog(@"[DEBUG] com.williamrijksen.onesignal load");
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onAppCreate:)
+                                                 name:@"UIApplicationDidFinishLaunchingNotification" object:nil];
+}
+
+- (void)handleQueuedBootEvent:(NSString*)eventName
+{
+    if (_queuedBootEvents != nil && [_queuedBootEvents objectForKey:eventName] != nil) {
+        NSLog(@"[DEBUG] com.williamrijksen.onesignal Fire event from startup %@", eventName);
+        [[NSNotificationCenter defaultCenter] postNotificationName:eventName object:self userInfo:[_queuedBootEvents objectForKey:eventName]];
+        [_queuedBootEvents removeObjectForKey:eventName];
+    }
+}
+
+-(void)_listenerAdded:(NSString*)type count:(int)count
+{
+    NSLog(@"[DEBUG] com.williamrijksen.onesignal add listener %@", type);
+    if (count == 1 && [type isEqual:TiNotificationOpened]) {
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(notificationOpened:) name:TiNotificationOpened object:nil];
+        [self handleQueuedBootEvent:type];
+    }
+    if (count == 1 && [type isEqual:TiNotificationReceived]) {
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(notificationReceived:) name:TiNotificationReceived object:nil];
+        [self handleQueuedBootEvent:type];
+    }
+}
+
+-(void)_listenerRemoved:(NSString*)type count:(int)count
+{
+    NSLog(@"[DEBUG] com.williamrijksen.onesignal remove listener");
+    if (count == 0 && [type isEqual:TiNotificationOpened]) {
+        [[NSNotificationCenter defaultCenter] removeObserver:self name:TiNotificationOpened object:nil];
+    }
+    if (count == 0 && [type isEqual:TiNotificationReceived]) {
+        [[NSNotificationCenter defaultCenter] removeObserver:self name:TiNotificationReceived object:nil];
+    }
+}
+
+-(void)notificationOpened:(NSNotification*)info
+{
+    if (_instance != nil) {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+            NSMutableDictionary *event = [[NSMutableDictionary alloc] initWithDictionary: @{
+                                                                                        @"title" : [[info userInfo] valueForKey:@"title"],
+                                                                                        @"body" : [[info userInfo] valueForKey:@"body"],
+                                                                                        }];
+            NSLog(@"[DEBUG] com.williamrijksen.onesignal FIRE notificationOpened: %@" , event);
+            [self fireEvent:TiNotificationOpened withObject:event];
```

Comment:
Check if the proxy has listeners first, e.g `if ([self _hasListener:fireEvent:TiNotificationOpened])`